packages/vitest/src/integrations/wait.ts

```diff
@@ -0,0 +1,93 @@
+import { getSafeTimers } from '@vitest/utils'
+
+// The waitFor function was inspired by https://github.com/testing-library/web-testing-library/pull/2
+
+export type WaitForCallback<T> = () => T | Promise<T>
+
+export interface WaitForOptions {
+  /**
+   * @description Time in ms between each check callback
+   * @default 50ms
+   */
+  interval?: number
+  /**
+   * @description Time in ms after which the throw a timeout error
+   * @default 1000ms
+   */
+  timeout?: number
+}
+
+function copyStackTrace(target: Error, source: Error) {
+  if (source.stack !== undefined)
+    target.stack = source.stack.replace(source.message, target.message)
+  return target
+}
+
+export function waitFor<T>(callback: WaitForCallback<T>, options: number | WaitForOptions = {}) {
+  const { setTimeout, setInterval, clearTimeout, clearInterval } = getSafeTimers()
+  const { interval = 50, timeout = 1000 } = typeof options === 'number' ? { timeout: options } : options
+  const STACK_TRACE_ERROR = new Error('STACK_TRACE_ERROR')
+
+  return new Promise<T>((resolve, reject) => {
+    let lastError: unknown
+    let promiseStatus: 'idle' | 'pending' | 'resolved' | 'rejected' = 'idle'
+    let timeoutId: ReturnType<typeof setTimeout>
+    let intervalId: ReturnType<typeof setInterval>
+
+    const onResolve = (result: T) => {
+      if (timeoutId)
+        clearTimeout(timeoutId)
+      if (intervalId)
+        clearInterval(intervalId)
+
+      resolve(result)
+    }
+
+    const handleTimeout = () => {
+      let error = lastError
+      if (!error)
+        error = copyStackTrace(new Error('Timed out in waitFor!'), STACK_TRACE_ERROR)
```

Comment:
Should we just add a `cause` to an error instead of patching it?

Reply:
This error needs stack because can help us find the call stack, If use `cause` instead of patching it the the error displays like it.

<img width="512" alt="image" src="https://github.com/vitest-dev/vitest/assets/29533304/faf161ed-bc89-4862-af36-6a36a26476c8">

The expected should be this

<img width="480" alt="image" src="https://github.com/vitest-dev/vitest/assets/29533304/723ad358-0c71-4b87-8d28-d125ba2f1678">

Reply:
The cause should be the other error, not timed-out one. If it can't show the preview then it's a bun in Vitest that we should fix instead of hide it like this.

Reply:
> The cause should be the user error, not timed-out one. 

What does mean "user error"?  The patching stack trace behaviour only executes when the callback doesn't throw any error. So we haven't `cause`, just timed out error.

> If it can't show the preview then it's a bug in Vitest that we should fix instead of hide it like this.

Seems like the `cause` doesn't show code frame is intentional.  Do you know why?

https://github.com/vitest-dev/vitest/blob/65b200f0dd44357d26eac740d7d097d364078bd8/packages/vitest/src/node/error.ts#L105-L107

Reply:
Hm, now I see what this code does. Ok, since we support different browsers, let's keep the patch.